File: 장세훈/LeetCode/medium/AddTwoNumbers.java
class AddTwoNumbers {
    /*
        2. Add Two Numbers
        https://leetcode.com/problems/add-two-numbers/description/

        숫자 범위가 너무 크기때문에 단순 덧셈으로 계산이 안됨
        전가산기 개념으로 carry, remainder 변수를 별도로 놓고 덧셈
     */
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode node = new ListNode(0);
        ListNode root = node;

        int sum, carry = 0, remainder;

        while (l1 != null || l2 != null || carry != 0) {
            sum = 0;
            if (l1 != null) {
                sum += l1.val;
                l1 = l1.next;
            }

            if (l2 != null) {
                sum += l2.val;
                l2 = l2.next;
            }

            remainder = (sum + carry) % 10;
            carry = (sum + carry) / 10;
            node.next = new ListNode(remainder);
            node = node.next;
        }

        return root.next;
    }
}

File: 정수빈/LeetCode/medium/add-two-numbers.py
# https://leetcode.com/problems/add-two-numbers/
# 설명: 링크드리스트의 값을 합해서 다시 링크드리스트로 반환하는 문제입니다. while을 사용해서 링크드리스트의 값을 
# 전부 가져온 다음, 합하고 링크드리스트를 만들었습니다. 
# 난이도: medium

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: Optional[ListNode]
        :type l2: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        node1 = l1
        node1_num = str(node1.val)
        node2 = l2
        node2_num = str(node2.val)
        while node1.next:
            node1_num = node1_num + str(node1.next.val)
            
            node1 = node1.next

        while node2.next:
            node2_num = node2_num + str(node2.next.val)
            
            node2 = node2.next

        node_num = str(int(node1_num[::-1]) + int(node2_num[::-1]))
        idx = len(node_num)-1
        node = ListNode(int(node_num[idx]))
        root = node
        while idx > 0:
            idx -= 1
            node.next = ListNode(int(node_num[idx]))
            node = node.next
            
        return root
        